🕊️ Migration `PPOv2` -> `PPO` (#2174)
huggingface/trl: docs/source/customization.mdx

Reviewer: Although we use the `DPOTrainer` in these examples, should we start the doc with a small comment that they customisation applies to most (all?) trainers?

Author: Done in 59515ca

Consult the 🤗 Accelerate [documentation](https://huggingface.co/docs/accelerate/usage_guides/deepspeed) for more information about the DeepSpeed plugin.

> ## Use different optimizers and schedulers

Reviewer: Lots of this doc come from the time when we just had `PPOTrainer` which wasn't a subclass of `transformers`. I think we should remove anything that can be achieved on the `transformers` side, perhaps with an opening comment like:

"Every trainer in TRL is a subclass of the `Trainer` from 🤗 Transformers. This means you can change the optimizer or learning rate scheduler by simply changing the corresponding attributes of the TRL trainer configuration class. See the 🤗 Transformers [documentation](https://huggingface.co/docs/transformers/perf_train_gpu_one#optimizer-choice) for more details."

Note we can also nuke the deepspeed doc above since that's a relic from the old PPO days

Author: Agree with you. Let's apply these recommendation in other PRs (see https://github.com/huggingface/trl/pull/2174#issuecomment-2399843454)

> By default, the `DPOTrainer` creates a `torch.optim.Adam` optimizer. You can create and define a different optimizer and pass it to `DPOTrainer`:

Reviewer: nit:
```suggestion
By default, the `DPOTrainer` creates a `torch.optim.AdamW` optimizer. You can create and define a different optimizer and pass it to `DPOTrainer` as follows:
```

Since we subclass `transformers.Trainer` nowadays, I'd be tempted to remove this part altogether since it's the same logic in `transformers`

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from torch import optim
from trl import DPOConfig, DPOTrainer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")
training_args = DPOConfig(output_dir="Qwen2.5-0.5B-DPO")

optimizer = optim.SGD(model.parameters(), lr=training_args.learning_rate)

trainer = DPOTrainer(
    model=model,
    args=training_args,
    train_dataset=dataset,
    tokenizer=tokenizer,
    optimizers=(optimizer, None),
)
trainer.train()
```

> ### Use 8-bit optimizer

Reviewer: Same here, I think this can be removed in favour of this: https://huggingface.co/docs/transformers/v4.45.2/en/perf_train_gpu_one#optimizer-choice

Author: Done in 9bfab24

For memory efficient fine-tuning, you can also pass `Adam8bit` optimizer from `bitsandbytes`:

```python
import bitsandbytes as bnb
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")
training_args = DPOConfig(output_dir="Qwen2.5-0.5B-DPO")

optimizer = bnb.optim.Adam8bit(model.parameters(), lr=training_args.learning_rate)

trainer = DPOTrainer(
    model=model,
    args=training_args,
    train_dataset=dataset,
    tokenizer=tokenizer,
    optimizers=(optimizer, None),
)
trainer.train()
```

### Use LION optimizer

> You can use the new [LION optimizer from Google](https://huggingface.co/papers/2302.06675) as well, first take the source code of the optimizer definition [here](https://github.com/lucidrains/lion-pytorch/blob/main/lion_pytorch/lion_pytorch.py), and copy it so that you can import the optimizer. Make sure to initialize the optimizer by considering the trainable parameters only for a more memory efficient training:

Author: done in f43896a

    optimizers=(optimizer, lr_scheduler),
)
trainer.train()
```

## Memory efficient fine-tuning by sharing layers

Another tool you can use for more memory efficient fine-tuning is to share layers between the reference model and the model you want to train.

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import create_reference_model, DPOConfig, DPOTrainer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
ref_model = create_reference_model(model, num_shared_layers=6)
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:1%]")
training_args = DPOConfig(output_dir="Qwen2.5-0.5B-DPO")

trainer = DPOTrainer(
    model=model,
    ref_model=ref_model,
    args=training_args,
    train_dataset=dataset,
    tokenizer=tokenizer,
)
trainer.train()
```

## Pass 8-bit reference models 
 
Since `trl` supports all keyword arguments when loading a model from `transformers` using `from_pretrained`, you can also leverage `load_in_8bit` from `transformers` for more memory efficient fine-tuning.

> Read more about 8-bit model loading in `transformers` [here](https://huggingface.co/docs/transformers/perf_infer_gpu_one#bitsandbytes-integration-for-int8-mixedprecision-matrix-decomposition).

Reviewer: ```suggestion
Read more about 8-bit model loading in `transformers` [here](https://huggingface.co/docs/transformers/v4.45.2/en/peft#load-in-8bit-or-4bit).
```

Author: ```suggestion
Read more about 8-bit model loading in `transformers` [here](https://huggingface.co/docs/transformers/en/peft#load-in-8bit-or-4bit).
```

```python
from datasets import load_dataset

> from transformers import AutoModelForCausalLM, AutoTokenizer

Reviewer: ```suggestion
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
```

from trl import DPOConfig, DPOTrainer

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")

> ref_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", load_in_8bit=True)

Reviewer: I think this is the preferred API now:

```suggestion
quantization_config = BitsAndBytesConfig(load_in_8bit=True)
ref_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", quantization_config= quantization_config)
```
